janusgraph-core/src/main/java/org/janusgraph/graphdb/query/QueryUtil.java

```diff
@@ -241,7 +262,30 @@ public static boolean isEmpty(Condition<?> condition) {
     private static <E extends JanusGraphElement> Or<E> addConstraint(final RelationType type, OrJanusPredicate predicate, List<Object> values, Or<E> or, StandardJanusGraphTx tx) {
         for (int i = 0 ; i < values.size(); i++) {
             final JanusGraphPredicate janusGraphPredicate = predicate.get(i);
-            if (janusGraphPredicate instanceof AndJanusPredicate) {
+            if (janusGraphPredicate instanceof Contain) {
+                //Rewrite contains conditions
+                final Collection childValues = (Collection) values.get(i);
+                if (janusGraphPredicate == Contain.NOT_IN) {
+                    if (childValues.size() == 1) {
+                        addConstraint(type, Cmp.NOT_EQUAL, childValues.iterator().next(), or, tx);
+                    }
+                    // Don't need to handle the case where childValues is empty, because it defaults to
+                    // an or(and()) is added, which is a tautology
+                    final And<E> nested = new And<>(childValues.size());
+                    for (final Object inValue : childValues) {
+                        addConstraint(type, Cmp.NOT_EQUAL, inValue, nested, tx);
+                    }
+                    or.add(nested);
```

Comment:
I wonder if there is any performance difference between:

`OR(NOT_EQUAL(x) AND NOT_EQUAL(y))`
v.s.
`OR(NOT_EQUAL(x)) OR(NOT_EQUAL(y))`?

Reply:
I don't quite get your point here. It looks like your second case is missing a logical operator between the two `OR`s. Anyway, I can't spot what kind of possible change in the code you refer to, in order to create one or the other of your examples.

Reply:
Oh never mind, I got it wrong